### PR TITLE
MVC TUI client

### PR DESCRIPTION
**Pull request by yuriykulikov**

### CODE REVIEW

File: tcp-client/src/controller.rs [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    let selected_channel_name = model.selected_channel_name();
+    let mut client = TcpStream::connect("127.0.0.1:8080").unwrap();
+
+    let message = match selected_channel_name.as_str() {
+        "+" => format!(
+            "PUBLISH {}: {}\n",
+            composed_email_content, "Created the channel"
+        ),
+        _ => format!(
+            "PUBLISH {}: {}\n",
```
Comment (tspaudi) [inline tcp-client/src/controller.rs:62]: Hm... would love to see some more fancyness in data serialisation :)

File: tcp-client/src/controller.rs [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    client.write_all(message.as_ref()).unwrap();
+
+    if selected_channel_name == "+" {
+        model.select_channel(composed_email_content.as_str());
+    }
+    model.composed_clear();
+}
+
+/// Emits key events and ticks at least every 200ms
+pub fn key_events() -> Receiver<Event<KeyEvent>> {
```
Comment (tspaudi) [inline tcp-client/src/controller.rs:76]: Well, nice way to read key events. Is there something which is offered by the RUST language OOTB?

File: tcp-server/src/main.rs [hunk 3]
```diff
@@ -2,38 +2,33 @@ use std::borrow::BorrowMut;
...
 struct VecDequeMailbox {
     data: Mutex<VecDeque<String>>,
 }
 
-impl Mailbox for VecDequeMailbox {
-    /// TODO is there a way (and does it make sense) to accept &str?
-    fn append(&self, email: String) {
-        let result = self.data.lock();
-        result.unwrap().push_front(email);
+impl VecDequeMailbox {
```
Comment (tspaudi) [inline tcp-server/src/main.rs:21]: Nice

File: tcp-server/src/main.rs [hunk 4]
```diff
@@ -50,47 +45,53 @@ impl VecDequeMailbox {
...
+#[tokio::main]
+async fn main() -> io::Result<()> {
+    tokio::join!(spawn_monitoring_thread(), spawn_tcp_listener(),);
     Ok(())
 }
 
 /// Spawns TcpListener thread, connects to 127.0.0.1:8080 and spawns thread per connection
-fn spawn_tcp_listener() {
-    thread::spawn(|| {
+fn spawn_tcp_listener() -> JoinHandle<()> {
```
Comment (tspaudi) [inline tcp-server/src/main.rs:55]: Nice way of doing it - a seperate tokio task which spawns more tasks. This resolves an issue I had. Thanks

File: tcp-server/src/main.rs [hunk 5]
```diff
@@ -50,47 +45,53 @@ impl VecDequeMailbox {
...
-fn handle_client(mut client: TcpStream, mailbox: Arc<Box<VecDequeMailbox>>) -> Result<(), io::Error> {
+async fn handle_client(
+    mut tcp_stream: BufReader<TcpStream>,
+    mailbox: Arc<Box<VecDequeMailbox>>,
+) -> Result<(), io::Error> {
     loop {
         let mut str = String::new();
-        BufReader::new(&client).read_line(&mut str);
+        let ret = tcp_stream.read_line(&mut str).await;
+        if let Ok(0) = ret {
```
Comment (tspaudi) [inline tcp-server/src/main.rs:78]: Magic Number.

**Review by tspopp — CHANGES_REQUESTED** [review]
- Please add tcp-server to your root Cargo.toml file, so I am able to run it with "cargo run --bin tcp-server".
- Please fix warnings / clippy of project tcp-server
- Feature Request: Cleanup Mailbox without restarting the server :laughing:

File: tcp-client/src/controller.rs (hunk 2 above)
Comment (yuriykulikov) [reply tcp-client/src/controller.rs:76]: I don't think so. Maybe I am wrong. Anyway, I want to replace this with a stream like Anatol suggested

File: tcp-server/src/main.rs (hunk 5 above)
Comment (yuriykulikov) [reply tcp-server/src/main.rs:78]: This is arguable. In many best practices 0 and 1 are not considered magic numbers. Here 0 means that 0 bytes were read indicating an EOF.

**Pull request merged**